crates/sui-core/src/checkpoints/casual_order.rs

```diff
@@ -64,38 +67,134 @@ impl CasualOrder {
     }
 }
 
+struct TransactionDependencies {
+    digest: TransactionDigest,
+    dependencies: BTreeSet<TransactionDigest>,
+    effects: TransactionEffects,
+}
+
+impl TransactionDependencies {
+    fn from_effects(effects: TransactionEffects, rwlock_builder: &RWLockDependencyBuilder) -> Self {
+        let mut dependencies: BTreeSet<_> = effects.dependencies.iter().cloned().collect();
+        rwlock_builder.add_dependencies_for(effects.transaction_digest, &mut dependencies);
+        Self {
+            digest: effects.transaction_digest,
+            dependencies,
+            effects,
+        }
+    }
+}
+
+/// Supplies TransactionDependencies tree with additional edges from transactions
+/// that write shared locks object to transactions that read previous version of this object.
+///
+/// With RWLocks we can have multiple transaction that depend on shared object version N - many read
+/// transactions and single write transaction. Those transactions depend on transaction that has written N,
+/// but they do not depend on each other. And specifically, transaction that reads N and writes N+1
+/// does not depend on read-only transactions that also read N.
+///
+/// We do not add such read transactions to TransactionEffects of shared object write transactions
+/// for next version to make sure TransactionEffects are not grow too large
+/// (and because you do not need read transactions to replay write transaction for next version).
+///
+/// However, when building checkpoints we supply transaction dependency tree with additional dependency edges to
+/// make it look like write transaction for next version casually depends on transactions that read
+/// previous versions, for two reasons:
+///
+/// (1) Without this addition we could have peculiar checkpoints where transaction reading
+/// version N appears after transaction that overwritten this object with version N+1.
+/// This does not affect how transaction is executed, but it is not something one would expect in
+/// casually ordered list.
+///
+/// (2) On the practical side it will allow to simplify pruner as it can now just tail checkpoints
+/// and delete objects in order they appear in TransactionEffects::modified_at_versions in checkpoint.
+struct RWLockDependencyBuilder {
+    read_version: HashMap<ObjectVer, Vec<TransactionDigest>>,
+    overwrite_versions: HashMap<TransactionDigest, Vec<ObjectVer>>,
+}
+
+type ObjectVer = (ObjectID, SequenceNumber);
+
+impl RWLockDependencyBuilder {
+    pub fn from_effects(effects: &[TransactionEffects]) -> Self {
```

Comment:
i think these methods don't need to be pub to be accessed by the tests

Reply:
I like to just keep methods as `pub` if they are accessed outside of struct. This struct itself is private, so otherwise pub or not on methods have no effects